### PR TITLE
MM-13844: Show rendered emojis in the text composition areas

### PR DESCRIPTION
**Pull request by ChrisDobby**
#### Summary
Adds a new component for text composition - components/post_editor.tsx - that displays emojis inline while composing a post.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10163

#### Screenshots
post_editor

### CODE REVIEW

**Comment by etoaster** [comment]
@hanzei I don't think I'm the right person. You may have been looking for @esethna?

**Comment by esethna** [comment]
Thanks for the PR! A few issues I noticed:

1. Short form emojis should not render until the space key is pressed, for example `:s`
2. Copying and pasting seems to copy the background:
image

3. From the example above, after hitting enter to post, the emoji is not included:
image

CC// QA Team on some cases to test ^

**Comment by esethna** [comment]
Hi @ChrisDobby, any questions on the feedback above?

**Comment by ChrisDobby** [comment]
> Hi @ChrisDobby, any questions on the feedback above?

Thanks @esethna - no questions as far as I can see.  Hopefully I'll update the PR sometime in the next couple of days

**Comment by esethna** [comment]
@ChrisDobby let us know if you need any help here :)

**Comment by ChrisDobby** [comment]
> @ChrisDobby let us know if you need any help here :)

Thanks @esethna - I've updated the PR so now copy and pasting will use the text rather than the html and when typing short form emojis they need to be followed by a space in the editor.

File: components/textbox/textbox.jsx [hunk 1]
```diff
@@ -242,7 +243,7 @@ export default class Textbox extends React.PureComponent {
                     onBlur={this.handleBlur}
                     onHeightChange={this.handleHeightChange}
                     style={{visibility: this.props.preview ? 'hidden' : 'visible'}}
-                    inputComponent={AutosizeTextarea}
+                    inputComponent={PostEditor}
```
Comment (Willyfrog) [inline components/textbox/textbox.jsx]: I'm worried that we are losing all custom code from the previous component, would this pass cypress or some of the QA tests for regression?

File: components/post_editor.tsx [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+        return (
+            <>
+                {placeholder && this.state.showingPlaceholder && (
+                    <div
+                        {...props}
+                        style={style.placeholder}
+                    >
+                        {placeholder}
+                    </div>)}
+                <div
```
Comment (Willyfrog) [inline components/post_editor.tsx]: I wonder if it wouldn't be easier to have the actual text someplace and generate the nodes from it, so when you do some operation you just have to do it on the text value

**Comment by hmhealey** [comment]
Wow, this sounds excellent. I'll have to set aside some time to take a thorough look through of this code since there's a lot going on here and I know the annoying parts of the autosizing text area really well.

**Comment by Willyfrog** [comment]
@ChrisDobby let me know if you have any questions with the feedback or any other part

File: components/post_editor.tsx [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+            return;
+        }
+
+        const {withEmoticons, tokens} = this.getEmoticons(value);
+        (this.refs.editor as Element).innerHTML = replaceTokens(withEmoticons, tokens);
+    }
+
+    private rearrangeNodes = (range: Range, currentNode: Node) => {
+        // If we've continued typing after creating an emoji then we need to remove the emoji span and
+        // add all the text into the last text node so we can check if there's still a valid emoji there
```
Comment (hmhealey) [inline components/post_editor.tsx]: I'm not sure if I understand this. Does this happen after a user has typed inside of an emoji's span? Maybe there's something I'm missing with what this function does?

**Comment by hmhealey** [comment]
One other thing, @esethna, we should put this behind a user-specific feature flag. I'm sure we can get this tested and working well on the PR branch, but we should put this as experimental to start while we let it soak on community-daily. The post textbox is one of the most important things in the app for users, and we'll get a lot of flak if everything isn't 100% correct with it (see the response to Slack's WSIWYG editor).

@ChrisDobby Would you be able to add a setting to the Advanced tab of the Account Settings to toggle between this and the previous `AutosizeTextarea`? I assume it can go under the `Preview Pre-Release Features` section, although @esethna can confirm where it should go.

**Comment by ChrisDobby** [comment]
Thanks for the feedback @hmhealey @Willyfrog - hopefully I'll get chance to look at it over the next couple of days or so

**Comment by esethna** [comment]
@ChrisDobby I discussed with Harrison, we don't think adding a flag is necessary as long as we get this merged early in a release cycle and do thorough testing on it with QA for various text input types. So you can disregard this comment from Harrison:

> @ChrisDobby Would you be able to add a setting to the Advanced tab of the Account Settings to toggle between this and the previous AutosizeTextarea? I assume it can go under the Preview Pre-Release Features section

**Comment by ChrisDobby** [comment]
Sorry I haven't had chance to look at this yet - been a bit hectic at home and haven't found the time :-(  With a bit of luck will get the time over this weekend though....

**Comment by hmhealey** [comment]
@ChrisDobby No worries at all. It's a hectic time right now for everyone. There's no rush on getting this in.

**Comment by esethna** [comment]
@ChrisDobby hope all is well, just curious if you need any help or guidance here?

**Comment by esethna** [comment]
@ChrisDobby just circling back on this, let us know if you're still working on this or if we should close the PR?

**Pull request closed without merging**